tools/config_values_conversions/main.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	dir, err := filepath.Abs(f)
+	if err != nil {
+		panic(err)
+	}
+
+	for i := 0; i < 3; i++ { // ../../
+		dir = filepath.Dir(dir)
+	}
+
+	return dir
+}
+
+func searchConversions(dir string, workDir string) ([]string, error) {
+	packagePaths := make([]string, 0)
+
+	err := filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
+		if f != nil && f.IsDir() {
+			if f.Name() == "internal" {
+				return filepath.SkipDir
+			}
+			if f.Name() == "testdata" {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+		if filepath.Ext(path) != ".go" {
+			return nil
+		}
+		if strings.HasSuffix(path, "_test.go") {
+			return nil
+		}
+
+		trimDir := workDir
+		packagePath := filepath.Join(
+			deckhouseGoModuleName,
+			filepath.Dir(
+				strings.TrimPrefix(path, trimDir),
+			),
+		)
+		packagePaths = append(packagePaths, packagePath)
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return packagePaths, nil
+}
+
+const deckhouseGoModuleName = "github.com/deckhouse/deckhouse/"
+
+func main() {
+	workDir := cwd()
+
+	var (
+		output string
+		stream = os.Stdout
+	)
+	flag.StringVar(&output, "output", "", "output files for generated code")
+	flag.Parse()
+
+	outputs := make([]string, 0)
+	outputPaths := strings.Split(output, ",")
+	for _, path := range outputPaths {
+		outputs = append(outputs, path)
+	}
+
+	if len(outputs) == 0 {
+		panic(fmt.Errorf("no outputs specified"))
+	}
+
+	var buf bytes.Buffer
+
+	var conversionPackages []string
+
+	conversionDirs := []string{
+		filepath.Join(workDir, "global-hooks/config-values-conversion"),
+	}
+
+	moduleDirs, err := filepath.Glob(filepath.Join(workDir, "modules/*/config-values-conversion"))
+	if err != nil {
+		panic(fmt.Errorf("search conversions for CE: %v", err))
+	}
+	conversionDirs = append(conversionDirs, moduleDirs...)
+
+	moduleDirs, err = filepath.Glob(filepath.Join(workDir, "ee/modules/*/config-values-conversion"))
+	if err != nil {
+		panic(fmt.Errorf("search conversions for EE: %v", err))
+	}
+	conversionDirs = append(conversionDirs, moduleDirs...)
+
+	moduleDirs, err = filepath.Glob(filepath.Join(workDir, "ee/fe/modules/*/config-values-conversion"))
+	if err != nil {
+		panic(fmt.Errorf("search conversions for FE: %v", err))
+	}
+	conversionDirs = append(conversionDirs, moduleDirs...)
+
+	for _, dir := range conversionDirs {
+		packagePaths, err := searchConversions(dir, workDir)
+		if err != nil {
+			panic(fmt.Errorf("search conversions: %v", err))
+		}
+		conversionPackages = append(conversionPackages, packagePaths...)
+	}
+
+	sort.Strings(conversionPackages)
+
+	t := template.New("registry")
+	t, err = t.Parse(registryTemplate)
+	if err != nil {
+		panic(fmt.Errorf("parse template: %v", err))
+	}
+
+	err = t.Execute(&buf, conversionPackages)
+	if err != nil {
+		panic(fmt.Errorf("execute template: %v", err))
+	}
+
+	bufBytes := buf.Bytes()
+	for _, path := range outputs {
+		var err error
+		dir := filepath.Dir(path)
+		err = os.MkdirAll(dir, 0o755)
```

Comment:
Hm, what does this permission mode means?

Reply:
Oh, this TODO is about values becomes empty after conversion. This case is handled by callers, e.g. getModuleConfig.
I'll remove this TODO.